src/software/embedded/robot_diagnostics_cli/embedded_communication.py

```diff
@@ -0,0 +1,148 @@
+from proto.import_all_protos import *
+from embedded_data import EmbeddedData
+from google.protobuf.message import Message
+from software.embedded.constants.py_constants import (get_estop_config, EstopMode)
+from rich.progress import track
+import software.python_bindings as tbots_cpp
+from software.py_constants import *
+import time
+
+
+class EmbeddedCommunication:
+    """Communication class for sending/executing protos with the robots"""
+    def __init__(self):
+        """Establishes multicast connections and e-stop information"""
+        self.primitive_set_sender = None
+        self.receive_robot_status = None
+        self.embedded_data = EmbeddedData()
+        self.channel_id = int(self.embedded_data.get_channel_id())
+        self.robot_id = int(self.embedded_data.get_robot_id())
+
+        self.sequence_number = 0
+        self.command_duration_seconds = 2.0
+        self.send_primitive_interval_s = 0.01
+
+        self.epoch_timestamp_seconds = 0
+        self.battery_voltage = 0
+        self.primitive_packet_loss_percentage = 0
+        self.primitive_executor_step_time_ms = 0
+
+        # total primitives for this robots
+        self.robot_primitives = {}
+
+        # Set the primitive of the local robot
+        self.robot_primitives[self.robot_id] = Primitive(stop=StopPrimitive())
+
+        # initialising the estop
+        self.estop_reader = None
+        self.should_send_stop = False
+        self.estop_mode, self.estop_path = (
+            get_estop_config(keyboard_estop=False, disable_communication=False))
+
+        # Always in PHYSICAL_ESTOP mode within CLI
+        try:
+            self.estop_reader = tbots_cpp.ThreadedEstopReader(
+                self.estop_path, 115200
+            )
+            self.__update_estop_state()
+        except Exception as e:
+            raise Exception(f"Invalid Estop found at location {self.estop_path} as {e}")
+
+    def epoch_time(self):
+        return self.epoch_timestamp_seconds
+
+    def __receive_robot_status(self, robot_status: Message) -> None:
+        """Updates the dynamic information with the new RobotStatus message.
+        :param robot_status: The incoming RobotStatus proto from the robot to read
+        """
+        self.epoch_timestamp_seconds = robot_status.time_sent.epoch_timestamp_seconds
+        self.battery_voltage = robot_status.power_status.battery_voltage
+        self.primitive_packet_loss_percentage = robot_status.network_status.primitive_packet_loss_percentage
+        self.primitive_executor_step_time_ms = robot_status.thunderloop_status.primitive_executor_step_time_ms
```

Comment:
have you considered keeping this state in `self.embedded_data` so that all the data is in one place?